Сдача проекта

Seetochka opened the pull request:
Реализованы функциональности:
- Общие фильмы
- Рекомендации
- Лента событий
- Отзывы
- Поиск

Реализованы фичи:
- Вывод самых популярных фильмов по жанру и годам
- Удаление фильмов и пользователей
- Добавление режиссёров в фильмы

heck99 committed:
feet: добавить удаление фильмов по id

heck99 committed:
feet: добавить удаление пользователей по id

heck99 committed:
Merge pull request #10 from Seetochka/add-user-remove-endpoint

feet: добавить удаление пользователей по id

heck99 committed:
feet: добавить удаление пользователей по id

heck99 committed:
feet: добавить Review и CRUD операции

Seetochka committed:
feat: добавить рейтинги и жанры

Seetochka committed:
test: добавить тесты на рейтинги и жанры

heck99 committed:
Merge pull request #12 from Seetochka/develop

Develop

heck99 committed:
feet: добавить лайки и дизлайки постам

heck99 committed:
Merge branch 'develop' into add-reviews

EvgeniyaKuzmina committed:
feat: — добавить новую сущность Director — отвечает за хранение информации о режиссёрах
— реализовать класс репозиторий для взаимодействия сущности режиссёра с БД (создание, изменение, удаление, изменения режиссёра)
— реализовать все этапы начина с БД до выдачи результата пользователю через контроллер.
— реализовать новый метод в репозитории для сущности Film позволяющий получить выборку фильму по части строки в названии фильма.
— начать изменения в сущности Film по добавлению в неё сущности Director и изменения соответствущих методов в репозитории

EvgeniyaKuzmina committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/models/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

heck99 committed:
feet: добавить тесты для отзывов

heck99 committed:
bugfix: утранить ошибку в именовании пакета

RobinsonKrusoe committed:
Commit#1 Добавить функциональность «Лента событий»

Kortni-BOO committed:
Общие фильмы

EvgeniyaKuzmina committed:
feat: — добавить новую сущность Director — отвечает за хранение информации о режиссёрах
— начать изменения в сущности Film по добавлению в неё сущности Director и изменения соответствущих методов в репозитории.
— реализовать класс репозиторий для взаимодействия сущности режиссёра с БД (создание, изменение, удаление, изменения режиссёра)
— реализовать все этапы начина с БД до выдачи результата пользователю через контроллер.
— реализовать новый метод в репозитории для сущности Film позволяющий получить выборку фильму по части строки в названии фильма.
— реализовать метод позволяющий найти фильмы по id режиссёра отсортированные по году выпуска или по количеству лайков

Kortni-BOO committed:
Починила тест

EvgeniyaKuzmina committed:
fix: удалить у поля directors класса Film аннотацию  NotNull

EvgeniyaKuzmina committed:
fix: добавить проверку если у фильма не указан режиссёр. Теперь возвращает null при получении фильма без режиссёра

EvgeniyaKuzmina committed:
fix: очистить данные таблицы для тестов

EvgeniyaKuzmina committed:
fix: удалить заккоментированные строки из файла

heck99 committed:
feat: добавить выбор популярных фильмов по годам и жанрам

Kortni-BOO committed:
Добавила db в gitignore

Kortni-BOO committed:
Delete db

RobinsonKrusoe committed:
Commit#2 Добавить интеграцию с функционалом "Отзывы"

heck99 committed:
bugfix: удалить ненужный класс и исправить опечатку

Seetochka committed:
feat: добавить функционал рекомендаций

Seetochka committed:
test: добавить тесты на удаление пользователя и фильма

heck99 committed:
bugfix: изменить правила удаленя записей из таблиц проверит тесты

EvgeniyaKuzmina committed:
stile:
— удалить закомментированные строки из файла и лишние импорты
— вынести строки по которым происходит поиск и выполнение запросов в отдельный класс констант
— вынести логику обработки и проверки данных в классы сервисов
— изменить выброс исключения, если введён неверный id режиссёра, или если пытаются удалить режиссёра который уже указан в фильмах.
— изменить стиль комментариев для всех методов
— изменить названия методов в общем стиле на save, find и delete
— добавить описание нового функционала в файл README

EvgeniyaKuzmina committed:
stile:
— раскомментировала код

heck99 committed:
feet: добавить описание задачи в readme

Seetochka committed:
refactor: добавить функционал в Readme и дорабовать получение рейтинга

heck99 committed:
feat: добавить описание в README.md

Seetochka committed:
Merge pull request #16 from Seetochka/add-most-populars

feat: добавить выбор популярных фильмов по годам и жанрам

heck99 committed:
Merge branch 'develop' into add-reviews

Seetochka committed:
Merge branch 'develop' of github.com:Seetochka/java-filmorate into add-recommendations

heck99 committed:
Merge branch 'develop' into add-remove-endpoint

# Conflicts:
#	README.md
#	src/main/resources/data.sql

Seetochka committed:
refactor: добавить появившиеся параметры для получение популярных фильмов

heck99 committed:
feat: исправить ошибки

heck99 committed:
style: привести файлы в порядок

heck99 committed:
feet: добавить описание в README.md

Kortni-BOO committed:
Внесены правки по ревью

Kortni-BOO committed:
Решаю конфликты

Kortni-BOO committed:
Merge branch 'develop' into add-common-films

Kortni-BOO committed:
Решаю конфликты

Kortni-BOO committed:
Я уже не знаю

Kortni-BOO committed:
Добавляю метод в репозиторий

RobinsonKrusoe committed:
fix: Исправить замечания review

Seetochka committed:
Merge pull request #11 from Seetochka/add-remove-endpoint

Add remove endpoint

RobinsonKrusoe committed:
fix: Исправить замечания review

Seetochka committed:
Merge pull request #13 from Seetochka/add-reviews

Add reviews

EvgeniyaKuzmina committed:
fix: изменить тип данных для id со String на int, который методы получают на вход в контроллерах и сервисах.
style: изменить названия методов по поиску с search на find

EvgeniyaKuzmina committed:
fix: изменить метод удаления режиссёра с типа String на void

EvgeniyaKuzmina committed:
fix: удалить свойство required у переменной пути в методе deleteDirector

EvgeniyaKuzmina committed:
Merge remote-tracking branch 'origin/develop' into add-search

# Conflicts:
#	README.md
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/resources/data.sql
#	src/main/resources/schema.sql

EvgeniyaKuzmina committed:
получила обновлённый код из ветки develop

Seetochka committed:
Merge branch 'develop' of github.com:Seetochka/java-filmorate into add-recommendations

Seetochka committed:
refactor: исправляет ошибки после актуализации

EvgeniyaKuzmina committed:
fix: удалить дублирующий участок кода

EvgeniyaKuzmina committed:
fix: удалить лишний комментарий в коде,
заменить во всех методах кроме byId уровень логирования на error

Seetochka committed:
Merge pull request #15 from Seetochka/add-search

Add search

Seetochka committed:
Merge branch 'develop' of github.com:Seetochka/java-filmorate into add-feed

Seetochka committed:
refactor: исправляет ошибки после актуализации

Seetochka committed:
Merge pull request #19 from Seetochka/add-feed

Add feed Добавлен функционал "Лента событий" (попытка №2)

Kortni-BOO committed:
Добавлен обработчик ошибок

Kortni-BOO committed:
Добавлен обработчик ошибок изменила

Seetochka committed:
Merge branch 'develop' of github.com:Seetochka/java-filmorate into add-recommendations

Seetochka committed:
refactor: исправляет форматирование и дорабатывает Readme

Seetochka committed:
Merge pull request #20 from Seetochka/add-recommendations

Add recommendations

Seetochka committed:
Merge branch 'develop' of github.com:Seetochka/java-filmorate into add-common-films

Seetochka committed:
refactor: исправляет тесты после актуализации

heck99 committed:
feet: изменить под тесты

heck99 committed:
feet: сделать костыли для тестов

Seetochka committed:
docs: доработать Readme

Seetochka committed:
Merge pull request #17 from Seetochka/add-common-films

Add common films

Seetochka committed:
Merge pull request #21 from Seetochka/add-reviews

feet: изменить под тесты

eugene-lenkevich commented on src/main/java/ru/yandex/practicum/filmorate/storage/event/EventDBStorage.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Collection;
+
+@Slf4j
+@Repository
```
А вот DirectorDbStorage имеет аннотацию @Component :)

PS не уверен про уместность именно аннотации @Repository, она относится к Spring Data и имеет особенности, в частности может трактоваться как JPA-репозиторий, а это не так. Но это не точно, повод задуматься, хотя в данном случае все работать будет, fyi

Русской доки не нашел)

![image](https://user-images.githubusercontent.com/94845100/173895282-ab6475a5-aa78-49e6-bd62-a76624550270.png)

Seetochka replied:
Исправили

eugene-lenkevich reviewed (changes requested):
Привет, Светлана, спасибо за PR

В целом все супер :)
Есть совсем небольшие замечания, посмотри комментарии ниже, пожалуйста

heck99 commented on src/main/java/ru/yandex/practicum/filmorate/services/ReviewService.java:
```diff
@@ -0,0 +1,201 @@
+            throw new IncorrectParameterException("count");
+        }
+
+        if (id.isPresent()) {
```
Если в данном методе использовать лямбду, то в выражении нужно выполнить 2 действия: 
1. проверить, что пользователь с таким id существует вызовом filmService.findById(id.get());, но этот метод выкидывает исключение(исключение проверяемое, поэтому оно требует обязательной проверки внутри лямбды, а мне нужно, чтобы если возникло исключение он просто прокинул его вверх(в сигнатуре метода указано, что метод выкидывает исключение
2. вызвать метод и вернуть значение, которое он вернёт с помощью return storage.findByFilmId(id.get(), count);
как я понимаю в лямбде вообще нельзя использовать return
Можно было бы передавать обёртку Integer id в метод и просто проверять if(!id == null), но так пользователям этого метода будет неочевидно, что передача null это нормальное поведение метода, поэтому я использовал именно Optional

eugene-lenkevich replied:
Понял, в этом случае все верно, из лямбды так просто не выкинешь исключение, текущий функциональный интерфейс нас в это ограничивает :)

Seetochka commented on src/main/java/ru/yandex/practicum/filmorate/services/DirectorService.java:
```diff
@@ -0,0 +1,86 @@
+     * Обновление данных режиссёра
+     */
+    public Director updateDirector(Director director) throws ModelNotFoundException {
+        findDirectorById(director.getId());
```
Метод `findDirectorById` уже внутри себя проверяет наличие режиссера, и если его нет выбрасывает `ModelNotFoundException`, который обрабатывается в `ErrorHandler`

eugene-lenkevich replied:
Имею в виду здесь director.getId(), это еще до findDirectorById  :)

Seetochka committed:
refactor: доработать формирование строки ошибок

RobinsonKrusoe committed:
fix: исправить замечания после review

eugene-lenkevich commented on src/main/java/ru/yandex/practicum/filmorate/services/ReviewService.java:
```diff
@@ -0,0 +1,201 @@
+        Review review = findById(reviewId);
+        userService.findById(userId);
+
+        if (status.isPresent()) {
```
Тут как минимум дублирование кода - два кусочка кода отличаются только тем что в одном случае идет +2, а в другом +1, далее value == status.get(), то есть внутренний if не нужен во втором случае :)
В общем тут дело не в Optional, от него вообще можно избавиться - например методом orElse(...)

heck99 replied:
Второй ты проверяет же, что статус false, а не то, что он есть, я понял, что можно заменить status.get() на value, но от ИФА не избавиться, потому что если лайк уже стоит, то мы не должны ставить ещё один

eugene-lenkevich replied:
На самом деле тут хотелось бы просто избавиться от нагромождения вложенных if-ов и optional-ов, всего то, а как именно не суть важно :)

Seetochka replied:
Исправили

heck99 committed:
feet: добавить методы Optional

EvgeniyaKuzmina committed:
style: — изменит название класса с Contstant на Constants
— изменить аннотацию @Component на аннотацию @Service в классе DirectorService
— добавить обработку id в сущности Director при обновлении режиссёра. Теперь, если не указан id выбрасывается исключение NullPointerException

Seetochka committed:
Merge pull request #23 from Seetochka/add-recommendations

Доработать формирование строки ошибок

Seetochka committed:
Merge pull request #24 from Seetochka/add-feed

Исправлены замечания после review

Seetochka commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -82,14 +84,29 @@ public int deleteLike(int filmId, int userId) {
 
+    @Override
+    public Collection<Film> findFilmsByTitleAndDirector(String query, String by) {
+        return null;
```
Удалили реализацию InMemory

Seetochka committed:
Merge pull request #26 from Seetochka/add-reviews

исправил замечания после ревью

Seetochka commented on src/main/java/ru/yandex/practicum/filmorate/services/ReviewService.java:
```diff
@@ -0,0 +1,201 @@
+
+        Optional<Boolean> status = likeReviewStorage.getStatus(reviewId, userId);
+
+        if (status.isPresent()) {
```
Исправили

Seetochka committed:
Merge pull request #28 from Seetochka/add-search

style: — изменит название класса с Contstant на Constants

Seetochka commented on src/main/java/ru/yandex/practicum/filmorate/constants/Constant.java:
```diff
@@ -0,0 +1,10 @@
+package ru.yandex.practicum.filmorate.constants;
+
+public class Constant {
```
Исправили

eugene-lenkevich reviewed (approved):
Все супер, спасибо!

PS как мне кажется немаловажный момент заключается в том, чтобы у команды формировался свой единый стиль. Это очень упрощает сопровождение проекта. В дальнейшем команды будут меняться, через какое то время не останется никого, кто начинал проект, но, за счет того что все в принципе сделано единообразно, будет гораздо проще потом во всем разбираться в отличие  от того, если бы разные куски кода были выполнены в разном стиле (но это конечно не  серебряная пуля, а лишь один из множества моментов  =).
